File: android/app/src/main/java/jaclib/memory/heap/NativeHeap.java
package jaclib.memory.heap;


public final class NativeHeap {

    public long peer;

    public final int a;

    public boolean b;

    public NativeHeap(int arg0) {
        this.a = arg0;
        this.allocateHeap(this.a);
        this.b = true;
    }

    public void allocateHeap(int arg0) {}

    @Override
    public synchronized void finalize() throws Throwable {
        super.finalize();
        this.b();
    }

    public synchronized void deallocateBuffer(int arg0) {}

    public synchronized int allocateBuffer(int arg0, boolean arg1) { return 0; }

    public void deallocateHeap() {}

    public NativeHeapBuffer a(int arg0, boolean arg1) {
        if (!this.b) {
            throw new IllegalStateException();
        }
        return new NativeHeapBuffer(this, this.allocateBuffer(arg0, arg1), arg0);
    }

    public synchronized void put(int arg0, byte[] arg1, int arg2, int arg3, int arg4) {}

    public synchronized void get(int arg0, byte[] arg1, int arg2, int arg3, int arg4) {}

    public synchronized boolean a() {
        return this.b;
    }

    public synchronized long getBufferAddress(int arg0) { return 0L; }

    public synchronized void b() {
        if (this.b) {
            this.deallocateHeap();
        }
        this.b = false;
    }
}

File: src/jvmMain/kotlin/jaclib/memory/heap/NativeHeap.kt
package jaclib.memory.heap


actual class NativeHeap actual constructor(/*private*/ actual val b: Int) {
    /*private*/ actual val peer: Long = 0

    /*private*/ actual var a: Boolean

    init {
        this.allocateHeap(this.b)
        this.a = true
    }

    /*private*/ actual external fun allocateHeap(arg0: Int)

    @Synchronized
    actual fun a(): Boolean {
        return this.a
    }

    actual fun a(arg0: Int, arg1: Boolean): NativeHeapBuffer {
        check(this.a)
        return NativeHeapBuffer(this, this.allocateBuffer(arg0, arg1), arg0)
    }

    @Synchronized
    @Throws(Throwable::class)
    actual fun finalize() {
//        super.finalize()
        this.b()
    }

    @Synchronized
    actual fun b() {
        if (this.a) {
            this.deallocateHeap()
        }
        this.a = false
    }

    @Synchronized
    /*private*/ actual external fun allocateBuffer(arg0: Int, arg1: Boolean): Int

    @Synchronized
    /*private*/ actual external fun get(arg0: Int, arg1: ByteArray?, arg2: Int, arg3: Int, arg4: Int)

    @Synchronized
    actual external fun put(arg0: Int, arg1: ByteArray?, arg2: Int, arg3: Int, arg4: Int)

    /*private*/ actual external fun deallocateHeap()

    @Synchronized
    actual external fun getBufferAddress(arg0: Int): Long

    @Synchronized
    actual external fun deallocateBuffer(arg0: Int)
}
